### ISSUE TITLE
Add typescript option: When returning from custom type guard function, validate that the type of the guarded variable is compatible with return type

### ISSUE THREAD

**Issue by ilyub** [microsoft/TypeScript#44103]
# Suggestion

## 🔍 Search Terms

validate custom type guard function return

## ✅ Viability Checklist

This proposal meats this goal:
Statically identify constructs that are likely to be errors

It is type checking proposal. So, it will not change emitted code.

## ⭐ Suggestion

I suggest:

1. To add the following new syntax:

Valid code:
```typescript
function guard(value: unknown): value is string {
  if (typeof value === "string") return value is string;
  // Here "value is string" is equivalent to true.

  return value is not string;
  // Here "value is not string" is equivalent to false.
}
```

Invalid code:
```typescript
function guard(value: unknown): value is string {
  if (typeof value === "object") return value is string;
  // Should raise error because value is NOT string, it is object after applying typeof operator.

  if (typeof value === "string") return value is not string;
  // Should raise error because value is ACTUALLY string after applying typeof operator.

  return value is object;
  // Should raise error because "value is object" is incompatible 
  // with "value is string" in the return type definition.
}
```

2. To add compiler option that will prohibit old syntax (i.e. "return true", "return false") and will impose new safe syntax inside type guards.

3. To add compiler mode that will check old syntax as well, e.g.:
```typescript
function guard(value: unknown): value is string | number {
  if (typeof value === "object") return true;
  // Should raise error with this new compile mode enabled.
}
```

## 📃 Motivating Example

```typescript
function f(value: unknown): string | number {
  if (typeof value === "object") return value;
  // Raises error: Type 'object | null' is not assignable to type 'string | number'
}

function guard(value: unknown): value is string | number {
  if (typeof value === "object") return true;
  // No error
}
```

## 💻 Use Cases

> What do you want to use this for?

This feature will help to catch incorrect custom type guards.

> What shortcomings exist with current approaches?

This feature will not effect old code. So I see no shortcomings.

**Comment by jcalz** [comment]
You're supposed to fill out the whole viability checklist, and:

>❌ This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)

is probably not met, because you want the compiler to take `return value is string` and emit `return true` and take `return value is not string` and emit `return false`.  Not sure if there's some way to reword the suggestion so that it doesn't do this (for example, maybe `return true as value is string` or `return false as value is not string` or something?)

---

Anyway, you can sort of kind of get something similar to what you're asking for with some helper functions:


```ts
const is = <T,>() => Object.assign((t: T) => true,
    { not<U>(u: Exclude<U, T>) { return false } })

const isString = is<string>();

function goodGuard(value: unknown): value is string {
    if (typeof value === "string") return isString(value); // okay
    return isString.not(value); // okay
}

function badGuard(value: unknown): value is string {
    if (typeof value === "object") return isString(value); // error!
    // Argument of type 'object | null' is not assignable to parameter of type 'string'.

    if (typeof value === "string") return isString.not(value); // error!
    // Argument of type 'string' is not assignable to parameter of type 'never'.

    if (typeof value !== "string") return isString(value); // error!
    // Argument of type 'unknown' is not assignable to parameter of type 'string'.

    throw new Error();
}
```

[Playground link to code](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAlhGBeGAeAKgGgHwAoCUy2MA8gEYBWApsFAHQCGEEcA5mLrlAFwzqFJiUAE4BXKpgBQMGTADeMMCCioAqnlG8AogA9gAG1EATKmsx9shBcKpRRwsDABmDfRCowAvl-yTJoSFgEAGURODBWZHgIVGhhcNY8fABuPydRMFo4cBhWEBAjAHFRBmEjXAA3V3FeDIBrJQB3MHxeKsMPBBg4hPlpWTgnGC4ATwAHKhAh9vFkJBQAIh6IhcIbOwdo0PiIyuqqFJgAeiOYEDqGEf6ZdftHELCIuiUoPY7Dk7OLq880jKycmQGEUSmU3jUYPUmi02vtot1HpE5Nd4ENRhMpjAZh55osQJQaFBVjBbpsHjtWOCDsljqcqMJhCBhABCFGfACCwlYogAtlQwLBMVBxh4AOT46i0GAAH0Uon0+lF8JeMCYLHYDDI+g8UBAMDGpQYfKg9LOQ2FExgouWrFFdD8slRwwtk2mcNxMCWiOJpPuEG2CWeyipHzpDKZrMdHK5vP5gvNIqtNqVXRVarYYE12pguv1huNpqFidFYCoFXpdodAzRLsx2JgzPmnptPtsdy2iJDNM+9MZLLZp053L5ArNOeLUJAzRTiDTzAzWZ1eoNwiNtkLCct1sRlZRUAAFozGooqMetOHhARUr8-EA)

**Comment by ilyub** [comment]
I see your point that this is not non-ECMAScript syntax, which I missed.

But there is part 3 of the suggestion:

Without changing any syntax add typescript option (e.g. checkGuardedVariableTypeOnReturn) that will check existing syntax:
```typescript
function guard(value: unknown): value is string | number {
  if (typeof value === "object") return true;
}
```
With checkGuardedVariableTypeOnReturn enabled it should show error: 
```
The type of guarded variable 'value' at the moment of return is not compatible with return type. 
'object | null' is not compatible with 'value is string | number'.
```

My point is that in most cases, when you write custom type guard you can narrow value type with built-in type guards so that it has correct type at the moment of return:
```typescript
function guard(value: unknown): value is string | number {
  if (typeof value === "string" || typeof value === "number") {
    // By this point value already has type string | number thanks to built-in type guards
    return true;
  }
}
```
I hardly imagine situation where you can't write custom type guard with such mode (checkGuardedVariableTypeOnReturn ) enabled. But if you can't or don't want you can always turn it off.

On the other hand, the gual of typescript is to make code type safe and when it comes to custom type guards typescript fails this goal.

**Comment by ilyub** [comment]
I have changed the title of the feature request because I see that suggesting extended syntax is not compatible with typescript policies.

So, I can only suggest stricter check option for existing syntax.

**Comment by MartinJohns** [comment]
Duplicate of #29980.

**Comment by RyanCavanaugh** [comment]
For context, the motivating use case for UDTGs is for when you're doing runtime analysis that TS doesn't intrinsically understand. Asking that TS check your analysis in that case is a non-starter since, nearly by definition, you're only writing the UDTG *because* TS was unable to check your work.
